driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -202,7 +202,12 @@ public Cluster getCluster() {
                         switch (rm.kind) {
                             case PREPARED:
                                 Responses.Result.Prepared pmsg = (Responses.Result.Prepared) rm;
-                                PreparedStatement stmt = DefaultPreparedStatement.fromMessage(pmsg, cluster, query, poolsState.keyspace);
+                                String keyspaceToUse = poolsState.keyspace;
+                                if (keyspace != null) {
+                                    // TODO: possibly check protocol version here and if not supported throw an exception.
+                                    keyspaceToUse = keyspace;
```

Comment:
will do

Reply:
One thing about this is that for this scenario to happen, the statement has to be successfully prepared by the server, which means the table and query criteria etc have to match for this query for the target keyspace, but not necessarily completely, so I agree we should still have this check.

Reply:
Just had a thought, we could handle this better by checking the poolState before sending the request to the server to ensure the keyspaces match ahead of time.  I think we should still check here too just in case for whatever reason the pool state changes during the request (which is unlikely).

Reply:
we could do this in general (not just for prepared statements) so at least we get the error before the query is sent and with a more appropriate error.

Reply:
👍  for the change.

As to:
> we could handle this better by checking the poolState before sending the request to the server to ensure the keyspaces match ahead of time

It's not possible if the query string uses a qualified table name, since we don't parse it.
```
session.execute("USE ks1");
// session's keyspace does not match the statement's keyspace, yet the query is valid
session.execute(new SimpleStatement("SELECT * FROM ks2.bar").setKeyspace(ks2));
```